extractor/filesystem/language/javascript/pnpmlock/extractor.go

```diff
@@ -0,0 +1,274 @@
+
+	return matches[1], matches[2]
+}
+
+func parsePnpmLock(lockfile pnpmLockfile) []*extractor.Inventory {
+	packages := make([]*extractor.Inventory, 0, len(lockfile.Packages))
+
+	for s, pkg := range lockfile.Packages {
+		name, version := extractPnpmPackageNameAndVersion(s, lockfile.Version)
+
+		// "name" is only present if it's not in the dependency path and takes
+		// priority over whatever name we think we've extracted (if any)
+		if pkg.Name != "" {
+			name = pkg.Name
+		}
+
+		// "version" is only present if it's not in the dependency path and takes
+		// priority over whatever version we think we've extracted (if any)
+		if pkg.Version != "" {
+			version = pkg.Version
+		}
+
+		if name == "" || version == "" {
+			continue
+		}
+
+		commit := pkg.Resolution.Commit
+
+		if strings.HasPrefix(pkg.Resolution.Tarball, "https://codeload.github.com") {
+			re := regexp.MustCompile(`https://codeload\.github\.com(?:/[\w-.]+){2}/tar\.gz/(\w+)$`)
+			matched := re.FindStringSubmatch(pkg.Resolution.Tarball)
+
+			if matched != nil {
+				commit = matched[1]
+			}
+		}
+
+		depGroups := []string{}
+		if pkg.Dev {
+			depGroups = append(depGroups, "dev")
+		}
+
+		packages = append(packages, &extractor.Inventory{
+			Name:    name,
+			Version: version,
+			SourceCode: &extractor.SourceCodeIdentifier{
+				Commit: commit,
+			},
+			Metadata: osv.DepGroupMetadata{
+				DepGroupVals: depGroups,
+			},
+		})
+	}
+
+	return packages
+}
+
+// Extractor extracts pnpm-lock.yaml files.
+type Extractor struct{}
+
+// Name of the extractor
+func (e Extractor) Name() string { return "javascript/pnpmlock" }
+
+// Version of the extractor
+func (e Extractor) Version() int { return 0 }
+
+// Requirements of the extractor.
+func (e Extractor) Requirements() *plugin.Capabilities { return &plugin.Capabilities{} }
+
+// FileRequired returns true if the specified file matches apk status file pattern.
+func (e Extractor) FileRequired(path string, fileInfo fs.FileInfo) bool {
+	return filepath.Base(path) == "pnpm-lock.yaml"
+}
+
+// Extract extracts packages from a pnpm-lock.yaml file passed through the scan input.
+func (e Extractor) Extract(ctx context.Context, input *filesystem.ScanInput) ([]*extractor.Inventory, error) {
+	var parsedLockfile *pnpmLockfile
+
+	err := yaml.NewDecoder(input.Reader).Decode(&parsedLockfile)
+
+	if err != nil && !errors.Is(err, io.EOF) {
+		return []*extractor.Inventory{}, fmt.Errorf("could not extract from %s: %w", input.Path, err)
+	}
+
+	// this will happen if the file is empty
+	if parsedLockfile == nil {
+		parsedLockfile = &pnpmLockfile{}
+	}
+
+	inventories := parsePnpmLock(*parsedLockfile)
+	for i := range inventories {
+		inventories[i].Locations = []string{input.Path}
+	}
+
+	return inventories, nil
+}
+
+// ToPURL converts an inventory created by this extractor into a PURL.
+func (e Extractor) ToPURL(i *extractor.Inventory) (*purl.PackageURL, error) {
+	return &purl.PackageURL{
+		Type:    purl.TypeNPM,
+		Name:    i.Name,
```

Comment:
Name must be lower cased according to the spec: https://github.com/package-url/purl-spec/blob/master/PURL-TYPES.rst#npm

Though it should already be lower cased.